### PR TITLE
Add Drone pipeline to Teleport repo

### LINKED ISSUE #3357: Keep Teleport base images updated.
Teleport images often have an old image and due to how caching works in Dockerfiles `apt-get update` is never run. Even though these vulnerabilities don't affect Teleport (they're often in unused system libraries), it's better if the image does not contain them.

Look into options to like `--no-cache` or `apt-get update && time` to make sure the latest patches are included in our images.

### PR DESCRIPTION
**Pull request by webvictim**
Fixes #3357 

This PR adds a number of build pipelines to Drone.

1) Automatically run linter, unit tests and integration tests in a Docker container.
-- This will run against every PR to `master` or `branch/*`, and every push to `master` or `branch/*`
-- We will get a Slack notification if tests fail on `master` so we can make sure we haven't broken that branch.

2) Add Docker container builds as detailed in https://github.com/gravitational/teleport/issues/3357
-- These run every night via a cron job scheduled inside Drone.

3) Add various Teleport builds.
-- linux/amd64 OSS
-- linux/amd64 Enterprise
-- linux/amd64 Enterprise FIPS
-- linux/amd64 CentOS 6 OSS
-- linux/amd64 CentOS 6 Enterprise
-- linux/amd64 CentOS 6 Enterprise FIPS
-- windows/amd64 tsh-only
-- These artifacts will be built and pushed to a staging S3 bucket whenever we push a tag to Teleport repo which matches the `v*` format, assuming the tests all pass correctly.
-- We will be able to 'promote' these to the public-facing clientbuilds download bucket via a future pipeline to replace the current release process.

4) Add automatic runs of `milv` against any PR/commit that touches the `docs` directory.

5) Automatically rebuild and push various teleport-buildbox Docker containers which are used throughout the pipeline.
-- For efficiency, these containers will only be rebuilt and pushed whenever we make a commit to `master`.
-- The way Jenkins currently does this is to rebuild the buildbox container used for tests at the start of every test run, but because the Jenkins box has a local Docker cache this is a no-op 99% of the time. This process can cause a certain amount of flip-flopping back and forth between layers, however, if your PR on Jenkins changes the Dockerfile while other PRs on Jenkins uses the version from `master`.
-- I think that the best way to fix this is probably to do a combination - build the buildbox containers and push them to the registry when we push to `master`, then pull this image down and use it as a cache when we rebuild from the Dockerfile at the beginning of each test run. It should essentially be a no-op when the Dockerfile is unchanged, while still allowing us to test Dockerfile changes correctly.
-- The flip-flopping won’t happen with Drone because every test run is completely isolated in its own Kubernetes pod.
-- As such, the Drone method will need to be refined to implement this in future.

This YAML file is a bit wordy - I wanted to use YAML anchors to cut down on the boilerplate/repeated code, but unfortunately the YAML spec defines every document broken by `---` as totally separate, so we can't make cross-references between them.

One way to fix this in future may be to use something like [jsonnet](https://docs.drone.io/pipeline/scripting/jsonnet/) instead to cut down on repeated code blocks.

The intention is to merge this PR and have it sit running in parallel with Jenkins for a while for adequate testing.

### CODE REVIEW

**Comment by webvictim** [comment]
Closes #3357

**Comment by webvictim** [comment]
I am too angry to rewrite it. Please just approve this.

File: .drone.yml
Comment (awly) [inline .drone.yml:41]: I'm probably paranoid, but this is another supply chain risk.
This container will have access to our AWS credentials. Do these only have RW access to one bucket?

File: .drone.yml
Comment (awly) [inline .drone.yml:967]: should this be `PRODUCTION_QUAYIO_DOCKER_USERNAME` and `PRODUCTION_QUAYIO_DOCKER_PASSWORD`?

File: .drone.yml
Comment (webvictim) [reply .drone.yml:41]: Yes - the AWS credentials which are available to pull requests only have access to the dronestorage “staging” bucket and nothing else. They’re limited by an IAM role.

The production credentials aren’t available to pull requests and can only be accessed by an authorised user running a promotion command from within Drone itself.

Comment (webvictim) [reply .drone.yml:41]: (pull requests will also be unsigned and so have to be approved by an admin before they run - we can look at the code and decline if anyone is trying to dump credentials to `stdout` or something)

File: .drone.yml
Comment (webvictim) [reply .drone.yml:967]: Which part?

The use of the production credentials here is correct as we’re pushing to the `quay.io/gravitational/teleport` Docker repo. This is done by the cron job within Drone so again, credentials can’t be accessed by a PR.

The use of `PLUGIN_USERNAME` and `PLUGIN_PASSWORD` is the way Drone expands variables passed in using the `settings:` element (basically prepend with ‘PLUGIN_`, convert the names to uppercase and spaces to underscores)

Comment (awly) [reply .drone.yml:967]: Ah, I was confused because the stages above pass username/pass via `environment`, not `settings`.
It'd be great to keep this consistent.

File: .drone.yml
Comment (awly) [reply .drone.yml:41]: I'm mostly concerned about someone compromising the `meltwater` dockerhub account.
The attack scenario here is pretty sophisticated:
- pwn dockerhub account
- upload a malicious `drone-cache` image
- this image exfiltrates our staging s3 credentials
- attacker uploads a backdored version of build artifacts to staging s3 bucket
- authorized user promotes this to production during release
- our users pull down promoted binaries and get pwned through a backdoor

The attacker would need to target teleport specifically. It also assumes we don't catch anything during release testing.

Since all of that is very unlikely, we can probably keep it as is.
Maybe in some distant future we can revisit our CI dependencies.

**Review by awly — APPROVED** [review]

File: .drone.yml
Comment (webvictim) [reply .drone.yml:41]: That’s a valid point. It is a very sophisticated and unlikely set of actions as you say, but possible nonetheless.

The code is public (https://github.com/meltwater/drone-cache) and licensed under Apache 2.0, so we could fork the repo, build the plugin and upload our own image to prevent this in future if we want.

**Comment by webvictim** [comment]
@russjones Could I get a rubber stamp please? I can't merge without it.

**Review by benarent — APPROVED** [review]
Nice, looking forward to our new drone overlords.

File: .drone.yml
Comment (russjones) [reply .drone.yml:41]: I'm not sure if this is a uncommon attack vector, haven't popular NPM modules been hit by a similar attack multiple times recently? I believe in those cases the authors gave up ownership and moved on to other things, but the underlying principal is the same: we should be very careful about from whom we accept external dependencies.

I think we can take two approaches:

1. We remove the build cache. I think this is a feasible alternative because cutting builds is not a frequent operation and even if we double or triple the time to cut a build, it won't slow our velocity down. We spend substantially more time on design and development than cutting builds.

2. We maintain our own fork of the image.

What do you guys think?

File: .drone.yml
Comment (russjones) [inline .drone.yml:23]: Can this come from a variable? Otherwise we'll have to bump the version of Go in many different places.

**Comment by russjones** [comment]
@webvictim Once all issues are resolved, please rebase and squash to make this 1 commit before merging to master.

File: .drone.yml
Comment (webvictim) [reply .drone.yml:23]: Sort of. The problem is that the `golang` container tags are in the `1.13.2` format but all our own `RUNTIME` references (which we use to tag our `buildbox` containers and pass into builds etc) are in `go1.13.2` format. I'll see whether I can do some string manipulation to get it right.

The best I can manage is that we define `RUNTIME` once at the top of each pipeline, which will be 7 or 8 copies throughout the whole file.

File: .drone.yml
Comment (webvictim) [reply .drone.yml:46]: To be clear - this is a cache for tests rather than builds. We always build artifacts clean from source and don't use any caching.

**Pros:**
- The cache does make iterative testing substantially faster; it cuts the time for feedback roughly in half - about 7-8 minutes per commit with caching versus maybe 15 minutes or so without caching.

**Cons:**
- Hypothetical attack vector as described (which could be mitigated by us self-building/hosting the image)
- We have to run the linter/unit tests/integration tests sequentially to be able to make use of caching, whereas if we disabled caching we could run all of these 3 in parallel and the time taken would be about equivalent to the longest running tests (which I think are currently integration tests at ~6/7 minutes all told, I haven't checked in a while though)
- Caching has diminishing returns as PRs get up into the hundreds of commits - for example, the cache for this PR is now up to about 650MB compressed and it's taking ~30 seconds to download/decompress the cache and ~3 minutes to recompress/upload it again at the end. However, it's very rare for us to have PRs with ~200 commits and the improvements in feedback time lower down are good for quick development and confidence.

Maintaining our own fork of the image would not be prohibitively difficult if we wanted to go that way. I'm happy to disable caching entirely if we think that's a better solution. The caching would also make tests on the master/release branches substantially faster as we'd effectively only be recompiling the deltas between each push rather than the entirely of Teleport each time.

File: .drone.yml
Comment (webvictim) [reply .drone.yml]: This is a no-can-do because Drone doesn't seem to allow you to use variables in image names for pipeline steps.

To be honest, changing the version of Go should be as simple as a global search-and-replace on this file to change `1.13.2` to `1.14.4`. If we were going to bump the version of Go then we'd need to change it throughout the repo with a global search and replace anyway so I don't see it causing any trouble.

File: .drone.yml
Comment (webvictim) [reply .drone.yml:1013]: In general `settings:` is intended to be used with Drone plugins (which are coded to know how to process the `PLUGIN_VARIABLE` format) whereas `environment:` is for arbitrary images which you're passing data to. They are fairly interchangable because you can just look at `${PLUGIN_VARIABLE}` rather than `${VARIABLE}` if you want.

I've changed `drone-cache` to use the `settings` format instead of `environment`.

**Comment by webvictim** [comment]
@russjones Github will do that for me with the "squash and merge" option which we always use for PRs anyway.

File: .drone.yml
Comment (webvictim) [reply .drone.yml]: It seems like the caching is actually doing more harm than good on this branch.

Caching on:
Screenshot 2020-06-24 17 29 18

Caching off:
Screenshot 2020-06-24 17 29 10

Comment (webvictim) [reply .drone.yml]: I'll just remove all the caching and then we don't have to worry.

**Comment by webvictim** [comment]
@russjones The caching issue should be resolved; please take another look.

**Comment by webvictim** [comment]
retest this please

**Review by russjones — APPROVED** [review]

**Pull request merged**
